### PR TITLE
Rework MPI[Single]Ports and add Tests

### PR DESCRIPTION
**Pull request by fsimonis**
Adding the MPISinglePorts tests revealed some problems, which required to finally migrate the tests, which revealed some issues in the acceleration and com packages.

This PR will:
* Add tests for MPISinglePorts similar to #624 
* Refactor MPISinglePorts to distinguish between direct and global communicators.
* Fix not properly closed ports in MPIPorts and MPISinglePorts
* Migrate the tests to the context-bases system #702 
* Add a bypass to SocketCommunication to prevent work
* Detect the MPI implementation in CMake
* Disable mpiports tests when OpenMPI was detected #746 
* Refactors the ownership of cyclic comms from MVQN into ParallelMatrixOperations

### CODE REVIEW

File: src/com/MPIPortsCommunication.cpp [hunk 1]
```diff
@@ -47,20 +47,21 @@ void MPIPortsCommunication::acceptConnection(std::string const &acceptorName,
...
     // Connection
     MPI_Comm communicator;
     MPI_Comm_accept(const_cast<char *>(_portName.c_str()), MPI_INFO_NULL, 0, MPI_COMM_SELF, &communicator);
     PRECICE_DEBUG("Accepted connection at " << _portName << " for peer " << peerCurrent);
 
+    // Which rank is requesting a connection?
     int requesterRank = -1;
-    // Exchange information to which rank I am connected and which communicator size on the other side
     MPI_Recv(&requesterRank, 1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
+    // Who big is the communicator of the requester
```
Comment (uekerman) [inline src/com/MPIPortsCommunication.cpp:61]:
```suggestion
    // How big is the communicator of the requester
```

File: src/com/MPIPortsCommunication.cpp [hunk 2]
```diff
@@ -89,7 +94,7 @@ void MPIPortsCommunication::acceptConnectionAsServer(std::string const &acceptor
                                                      int                requesterCommunicatorSize)
 {
   PRECICE_TRACE(acceptorName, requesterName, acceptorRank, requesterCommunicatorSize);
-  PRECICE_CHECK(requesterCommunicatorSize > 0, "Requester communicator size has to be > 0!");
+  PRECICE_ASSERT(requesterCommunicatorSize >= 0, "Requester communicator size has to be positive!");
```
Comment (uekerman) [inline src/com/MPIPortsCommunication.cpp:97]:
```suggestion
  PRECICE_ASSERT(requesterCommunicatorSize >= 0, "Requester communicator size has to be positive.");
```

File: src/com/MPISinglePortsCommunication.cpp [hunk 3]
```diff
@@ -95,21 +101,30 @@ void MPISinglePortsCommunication::acceptConnectionAsServer(std::string const &ac
...
 
-  ConnectionInfoWriter conInfo(acceptorName, requesterName, tag, _addressDirectory);
-
   _isAcceptor = true;
 
-  if (utils::MasterSlave::getRank() == 0) { // only master opens a port
+  const int rank = utils::Parallel::current()->rank();
+
+  ///@todo verify that getRank() makes sense in this case
+  //if (utils::MasterSlave::getRank() == 0) { // only master opens a port
```
Comment (uekerman) [inline src/com/MPISinglePortsCommunication.cpp:109]: Can we clean up that?

File: src/com/MPISinglePortsCommunication.hpp [hunk 4]
```diff
@@ -76,8 +77,14 @@ class MPISinglePortsCommunication : public MPICommunication {
 
   std::string _addressDirectory;
 
-  /// Remote rank -> communicator map
-  std::map<int, MPI_Comm> _communicators;
+  /// A map of direct communication channels based on MPI_COMM_SELF on both sides
+  std::map<int, MPI_Comm> _direct;
+
+  /// The global intercommunicator that connects all ranks
+  MPI_Comm _global = MPI_COMM_NULL;
```
Comment (uekerman) [inline src/com/MPISinglePortsCommunication.hpp:84]: Why again do we need `_direct` **and** `_global` ?

File: src/com/SocketCommunication.cpp [hunk 5]
```diff
@@ -134,9 +134,15 @@ void SocketCommunication::acceptConnectionAsServer(std::string const &acceptorNa
                                                    int                requesterCommunicatorSize)
 {
   PRECICE_TRACE(acceptorName, requesterName, acceptorRank, requesterCommunicatorSize);
-  PRECICE_CHECK(requesterCommunicatorSize > 0, "Requester communicator size has to be > 0!");
+  PRECICE_CHECK(requesterCommunicatorSize >= 0, "Requester communicator size has to be positve!");
```
Comment (uekerman) [inline src/com/SocketCommunication.cpp:137]:
```suggestion
  PRECICE_ASSERT(requesterCommunicatorSize >= 0, "Requester communicator size has to be positve.");
```

**Comment by fsimonis** [comment]
> Wouldn't it be a nice add-on to print a warning if OpenMPI is used and MPIPortsCom or MPISinglePortsCom is chosen in CommunicationConfiguration?

True, but I would prefer to split this into a separate PR. This requires to pass the information from CMake to the configuration, which requires some additional work.

> Does changing the default as in #624 work for you now? Not saying that we should do this, but would be good to know. I cannot test this locally since OpenMPI.

I haven't tried #624 in conjunction with this PR yet, but will give it a try prior to merge.
You can always build the source with spack using:
```
$ cd ~/dev/veryimportant/preCICE
$ spack dev-build --until=build precice@develop^mpich
$ cd spack-build
$ spack build-env precice@develop^mpich -- ctest
```

File: src/com/MPISinglePortsCommunication.hpp (hunk 4 above)
Comment (fsimonis) [reply src/com/MPISinglePortsCommunication.hpp:84]: The `_direct` communication covers the case where P2P communication is required, namely:

* Master-Salves communication
* Master-Master communication

Once `acceptConnectionAsServer()` and `requestConnectionAsClient()` are whilst establishing the client-server (aka p2p, m2n) communication, this class connects both the client and the server communicator.
Once that happened, this `_global` communicator can be used for arbitrary p2p communication as well as collectives.

File: src/com/SocketCommunication.cpp (hunk 5 above)
Comment (fsimonis) [reply src/com/SocketCommunication.cpp:137]: This is called on each rank, thus this assertion must hold.

File: src/com/MPISinglePortsCommunication.cpp (hunk 3 above)
Comment (fsimonis) [inline src/com/MPISinglePortsCommunication.cpp:109]:
```suggestion
```

Yes we can :wink:

File: src/acceleration/MVQNAcceleration.cpp [hunk 6]
```diff
@@ -1,3 +1,4 @@
+#include "utils/assertion.hpp"
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/acceleration/MVQNAcceleration.cpp:1]:
```suggestion
```

You are right. I moved the requiring part to MatrixOperations

File: src/com/MPISinglePortsCommunication.hpp (hunk 4 above)
Comment (uekerman) [reply src/com/MPISinglePortsCommunication.hpp:84]: :+1: I did not fully get this from the comments above though. Could you please add this?

File: src/com/SocketCommunication.cpp (hunk 5 above)
Comment (uekerman) [reply src/com/SocketCommunication.cpp:137]: I am not sure if I understand what you mean. You mean that this should indeed be a `CHECK` and no `ASSERT` ?

Comment (fsimonis) [reply src/com/SocketCommunication.cpp:137]: The requester size always has to be `>0` as at least the Master calls this.
The requester size has to be checked by the remote SolverInterface, thus this should never be triggered. Therefore I think this should be an assertion.

Comment (uekerman) [reply src/com/SocketCommunication.cpp:137]: Yes, so the suggestion is good, right? :smirk:

Comment (fsimonis) [reply src/com/SocketCommunication.cpp:137]: Sorry! I misread the diff! :sweat_smile: 
Yes the suggestion is correct.

**Pull request merged**